src/epic/distances/divergence_free.py

```diff
@@ -0,0 +1,204 @@
+        function of the same equivalence class.
+
+        This is done by fitting a potential function (constructed from a neural
+        network) to minimize the L2 norm of the shaped reward function.
+        """
+        rew_fn = utils.multidim_rew_fn(reward_function)
+        n_samples_can = n_samples_can or self.default_samples_can
+        assert isinstance(n_samples_can, int)
+
+        ff_dim = max(self.state_dim * 4, 128)
+
+        net = nn.Sequential(
+            *[
+                nn.Flatten(),
+                nn.Linear(self.state_dim, ff_dim),
+                modules.Residual(
+                    nn.Sequential(
+                        nn.GELU(),
+                        nn.Linear(ff_dim, ff_dim),
+                        nn.GELU(),
+                    ),
+                ),
+                nn.Linear(ff_dim, 1),
+            ],
+        )
+
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        net.to(device)
+
+        max_epochs = 7500
+        optimizer = optim.AdamW(net.parameters(), lr=6e-4)
+        scheduler = optim.lr_scheduler.LambdaLR(
+            optimizer,
+            lambda epoch: 0.5 if (epoch > 5000 and epoch < 7500) else 0.25 if (epoch > 7500) else 1.0,
+        )
+
+        state_sample, action_sample, next_state_sample, done_sample = self.coverage_sampler.sample(
+            n_samples_can,
+        )
+
+        losses = []
+
+        def canonical_reward_fn(
+            state, action, next_state, done, /, return_tensor: bool = False, device: Union[str, torch.device] = "cpu"
+        ):
+            """Divergence-Free canonical reward function.
+
+            Args:
+                state: The batch of state samples from the coverage distribution.
+                action: The batch of action samples from the coverage distribution.
+                next_state: The batch of next state samples from the coverage distribution.
+                done: The batch of done samples from the coverage distribution.
+                return_tensor: Whether to return a torch.Tensor or a numpy nd.array.
+                device: The device on which to conduct computations.
+
+            Returns:
+                The canonicalized reward function.
+            """
+            n_samples_cov = state.shape[0]
+            assert n_samples_cov == action.shape[0] == next_state.shape[0] == done.shape[0]
+
+            state_tensor = utils.float_tensor_from_numpy(state, device)
+            next_state_tensor = utils.float_tensor_from_numpy(next_state, device)
+            if state_tensor.ndim == 1:
+                assert next_state_tensor.ndim == 1
+                state_tensor.unsqueeze_(-1)
+                next_state_tensor.unsqueeze_(-1)
+
+            net.to(device)
+
+            shaping = (self.discount_factor * net(next_state_tensor) - net(state_tensor)).squeeze(-1)
+
+            if not return_tensor:
+                return rew_fn(state, action, next_state, done) + utils.numpy_from_tensor(shaping)
```

Comment:
Worth adding an `assert` that no broadcasting is happening here (i.e. both summands have the same `ndim`). E.g. if someone passes in a neural net as the reward function, and the neural net outputs a singleton dimension at the end, pytorch is going to add a singleton dimension to the front of the potential to match the number of dimensions, and then broadcast (which leads to the wrong result). Can be an incredibly annoying bug to find, so good to be defensive with shape asserts

Reply:
Done.